### PR TITLE
[google_maps_flutter_web] Options to disable tilt controls and configure gesture handling

### PR DESCRIPTION
**Pull request by Rexios80**
Transferred from https://github.com/flutter/plugins/pull/4916

tiltControlsEnabled: On web, in satellite view, and at close zoom levels, the map displays tilt controls. This lets you disable them.

gestureHandling: On web, without greedy gesture handling the map can have an overlay saying you can only move the map with two fingers or Cmd+drag. This is not ideal, and greedy gestures lets you make that not happen. I added the other gesture handling options from the [Google Maps documentation](https://developers.google.com/maps/documentation/javascript/reference/map#MapOptions.gestureHandling) for completeness.

*List which issues are fixed by this PR. You must list at least one issue.*

[flutter/flutter/#99044](https://github.com/flutter/flutter/issues/99044)

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by reidbaker** [comment]
Thank you for your contribution. It looks like the tests are failing. Before we review the PR, please see what you can do to resolve the test failures. If you are unsure how to proceed, please reach out for help on the #hackers-new channel.

**Comment by Rexios80** [comment]
@reidbaker The actions seem to only be failing due to the required dependency overrides since this PR requires changes to the platform interface. Until we are ready to split out the platform interface changes into their own PR and get that merged, the actions will continue to fail on this PR.

File: packages/google_maps_flutter/google_maps_flutter/example/pubspec.yaml
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/example/pubspec.yaml:39]: FYI: If you add the following comment above this line:
```
# FOR TESTING ONLY. DO NOT MERGE.
```

(as the tooling that adds these overrides does) in this and all the other files, `pathified_analyze` will no longer report a failure.

File: packages/google_maps_flutter/google_maps_flutter/example/android/app/build.gradle
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/example/android/app/build.gradle:66]: Why did you change this?

File: packages/google_maps_flutter/google_maps_flutter/example/android/build.gradle
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/example/android/build.gradle:8]: And in general, why are you changing Android example configuration in this PR at all?

File: packages/google_maps_flutter/google_maps_flutter/lib/google_maps_flutter.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/lib/google_maps_flutter.dart:31]: Please keep the alphabetical order that is currently followed.

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:108]: If we're going to add multiple web-only settings, perhaps we should have a new `WebSettings` class to encapsulate them, so this method doesn't continue growing without bounds. (In general, it seems like we'd benefit from better organization of params here, so we may as well start now.)

The main reason not to do that would be if we believe this would ever be a non-web option. Is there any indication about that in the APIs?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/map_configuration.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/map_configuration.dart:26]: If we make the setting structure part of the platform interface package, we can use it here as well.

**Review by stuartmorgan — COMMENTED** [review]
Minor comments, but generally this looks good to be at an API level. I'll defer to @ditman on whether we have any better options in terms of how to expose this, but it seemed from the discussion in the previous iteration like we probably can't abstract it.

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:108]: Here is our original discussion on this issue: https://github.com/flutter/plugins/pull/4916#discussion_r879941100

I'm not sure we ever came to the conclusion that this option would or would not exist on other platforms

File: packages/google_maps_flutter/google_maps_flutter/example/pubspec.yaml
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/example/pubspec.yaml:40]: Regnerated these with the tooling

File: packages/google_maps_flutter/google_maps_flutter/example/android/app/build.gradle
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter/example/android/app/build.gradle]: Moved to https://github.com/flutter/packages/pull/3408

**Comment by Rexios80** [comment]
I think these checks are as good as they're getting as long as we have dependency overrides

**Comment by stuartmorgan** [comment]
Update from triage: This is just waiting for @ditman to have cycles to review it.

**Comment by stuartmorgan** [comment]
Update from triage: This is still waiting for @ditman to have bandwidth to review maps PRs.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:74]: I think that if the user is setting `scrollGesturesEnabled` or `zoomGesturesEnabled` to false, we should at least warn them that they now need to set `webGestureHandling` to some of its values instead (like "none"). Otherwise, we're just setting gestureHandling to `auto` when they possibly wanted `none`.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart:166]: This might have unintended consequences now that `mapID` and possibly vector tiles are coming. But we'll see when that happens if this is a problem :)

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/web_gesture_handling.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/web_gesture_handling.dart:6]: Have you considered making this an "enhanced enum" backed by a `String`? That should cut down in the use of `.name` below, right?

**Review by ditman — COMMENTED** [review]
All mobile implementations use webgl/vector maps to render stuff (and allow for arbitrary rotation of the map, zoom/tilt, etc...)

* 45 degree imagery seems to be only available in raster (*non-vector*) maps, in select locations (like San Francisco or Tokyo)
  * It is the official name [of the feature](https://developers.google.com/maps/documentation/javascript/examples/aerial-simple), so good for SEO I guess :)
* Vector (webgl) maps are coming once `mapID` support lands. See [docs](https://developers.google.com/maps/documentation/javascript/webgl).

I hope that in the future we won't need any more special web-only config parameters (or that they'll map better to their mobile counterparts if only because of the rendering engines are more similar!)

I think this can be landed with minimal tweaks?

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:74]: How should we warn them?

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart:166]: So are we leaving this as-is for now?

**Comment by stuartmorgan** [comment]
@ditman It looks like this is ready for another look.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:64]: Through the js `console` when in debug mode, like here:

* https://github.com/flutter/packages/blob/4c1149744dbb36984ec5e19a61443d91b7ecefb6/packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart#L222-L227

google_maps_flutter_web is using `dart:html` already, so you should be able to use this JS console object:

* https://api.flutter.dev/flutter/dart-html/Console/warn.html (`window.console.warn`)

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart:174]: Yes!

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/web_gesture_handling.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/web_gesture_handling.dart:6]: Enhanced enums let you override a `toString()` method that ends up used as expected in the JS code, see [this](https://github.com/flutter/packages/blob/main/packages/google_identity_services_web/lib/src/js_interop/shared.dart#L23-L40) as an example.

Here's some [Enhanced Enums Docs](https://dart.dev/language/enums#declaring-enhanced-enums).

But it's all right, let's leave this one as is.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:64]: Something like this:

```suggestion
  if (configuration.scrollGesturesEnabled == false ||
      configuration.zoomGesturesEnabled == false) {
    html.window.console.warn('`scrollGesturesEnabled` and `zoomGesturesEnabled` are now ignored on the web. Use `webGestureHandling: WebGestureHandling.none` instead.');
  }

  if (configuration.webGestureHandling != null) {
    options.gestureHandling = configuration.webGestureHandling!.name;
```

**Comment by ditman** [comment]
Let's release the platform_interface changes, right? @stuartmorgan  My other comments can be applied when the web package is ready for review!

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/web_gesture_handling.dart
Comment (Rexios80) [reply packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/web_gesture_handling.dart:6]: Right it is possible that way you just worded it weird 🙃 
If you're fine leaving it that's fine with me

**Comment by stuartmorgan** [comment]
> Let's release the platform_interface changes, right? @stuartmorgan My other comments can be applied when the web package is ready for review!

If you feel good about the state of the overall PR modulo small tweaks, then yes. @Rexios80 It sounds like we're ready for a sub-PR for the platform interface part!

**Comment by Rexios80** [comment]
@ditman @stuartmorgan #4478

**Comment by Rexios80** [comment]
The actions that are failing are pulling an old version of the platform interface. Is there something I'm supposed to do to fix that?

**Comment by stuartmorgan** [comment]
> The actions that are failing are pulling an old version of the platform interface. Is there something I'm supposed to do to fix that?

See the `analyze_downgraded` section of https://github.com/flutter/flutter/wiki/Understanding-Packages-tests#specific-tests

File: packages/google_maps_flutter/google_maps_flutter/pubspec.yaml
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter/pubspec.yaml:27]: Ideally, web should be released first, and then this dependency bumped to `^0.5.2`.

In practice, our tooling should be able to support this, but users may end up in an inconsistent state (that they can fix with a `flutter clean`, however).

WDYT @stuartmorgan, can this two packages be released at the same time? (Should they?)

Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter/pubspec.yaml:27]: They should not be; the implementation should be split out as another PR so that we can do the version update. We always want to do that for new features so that we don't have the app-facing package updating saying "You can now do X!" and then when people update only the app-facing package (which is common) trying to do X doesn't actually work. (We forget sometimes, and we actually just had a bug where we forgot and a client got an unimplemented exception when trying to use a new feature.)

There are some non-feature cases where we can combine all the non-interface packages into one commit, but for features it's usually a chain of three, not two.

**Comment by Rexios80** [comment]
@ditman @stuartmorgan Can we merge this now?

**Review by stuartmorgan — APPROVED** [review]
LGTM

This just needs a quick final stamp from @ditman to land.

**Comment by ditman** [comment]
> I hope that in the future we won't need any more special web-only config parameters (or that they'll map better to their mobile counterparts if only because of the rendering engines are more similar!)

Here's some more web-only parameters incoming: https://github.com/flutter/packages/pull/4553

Before we merge the web implementation of this, let me steer 4553 to create a new "webOptions" object or similar, where we add the 3 options proposed in 4553, and add the 45deg imagery and gesture handling from here, then we'll update the web implementation to read from the new "webOptions" object, instead of directly from the `MapConfiguration`.

/cc @stuartmorgan

**Comment by ditman** [comment]
(Discussed on chat with @stuartmorgan and we decided to go with the PRs as they are, and then maybe do a more thorough refactor of per-platform settings, for example "liteMode" doesn't make sense in the web or iOS either)

**Review by ditman — APPROVED** [review]
LGTM, let's go!

**Pull request merged**
